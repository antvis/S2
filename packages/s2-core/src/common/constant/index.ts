export * from './events';
export * from './basic';
export * from './classnames';
export * from './condition';
export * from './dataConfig';
export * from './frozen';
export * from './interaction';
export * from './options';
export * from './scroll';
export * from './theme';
export * from './tooltip';
export * from './resize';
