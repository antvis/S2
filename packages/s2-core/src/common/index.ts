export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
