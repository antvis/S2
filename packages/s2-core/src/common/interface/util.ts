export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
