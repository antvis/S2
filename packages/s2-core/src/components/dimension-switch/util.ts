import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
