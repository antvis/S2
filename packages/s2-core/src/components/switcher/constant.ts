import { ColIcon, RowIcon, ValueIcon } from '../icons/index';

export enum FieldType {
  Rows = 'rows',
  Cols = 'cols',
  Values = 'values',
}

export const SWITCHER_CONFIG = {
  [FieldType.Rows]: {
    text: '行头',
    icon: RowIcon,
  },
  [FieldType.Cols]: {
    text: '列头',
    icon: ColIcon,
  },
  [FieldType.Values]: {
    text: '值',
    icon: ValueIcon,
  },
} as const;
