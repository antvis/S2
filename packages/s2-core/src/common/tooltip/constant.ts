import { PRE_CLASS } from '@/common/constant/classnames';

export const TOOLTIP_CLASS_PRE = `${PRE_CLASS}-tooltip`;

export const TOOLTIP_OPERATION_CLASS_PRE = `${TOOLTIP_CLASS_PRE}-operation`;

export const DEFAULT_ICON_PROPS = {
  width: 14,
  height: 14,
  style: {
    verticalAlign: 'sub',
    marginRight: 4,
  },
};

export const POSITION_X_OFFSET = 10;
export const POSITION_Y_OFFSET = 10;
