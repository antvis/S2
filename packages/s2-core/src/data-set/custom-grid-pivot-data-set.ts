import { EXTRA_FIELD, i18n } from '../common';
import type { S2DataConfig } from '../common/interface';
import { CustomTreePivotDataSet } from './custom-tree-pivot-data-set';

export class CustomGridPivotDataSet extends CustomTreePivotDataSet {
  processDataCfg(dataCfg: S2DataConfig): S2DataConfig {
    const valueInCols = !this.spreadsheet.isCustomRowFields();
    const rows = valueInCols
      ? [EXTRA_FIELD]
      : [...(dataCfg.fields.rows || []), EXTRA_FIELD];
    const meta = this.processMeta(dataCfg.meta!, i18n('数值'));

    return {
      ...dataCfg,
      meta,
      fields: {
        ...dataCfg.fields,
        rows,
        valueInCols,
      },
    };
  }
}
