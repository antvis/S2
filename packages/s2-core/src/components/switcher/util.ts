import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
