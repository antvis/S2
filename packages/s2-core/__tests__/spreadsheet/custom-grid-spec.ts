import type { Group } from '@antv/g';
import { pick } from 'lodash';
import { CustomGridData } from 'tests/data/data-custom-grid';
import { waitForRender } from 'tests/util';
import { getContainer } from 'tests/util/helpers';
import { KEY_GROUP_COL_RESIZE_AREA } from '../../src/common/constant';
import { CustomGridPivotDataSet } from '../../src/data-set/custom-grid-pivot-data-set';
import {
  customColGridSimpleFields,
  customRowGridSimpleFields,
} from '../data/custom-grid-simple-fields';
import {
  expectHighlightActiveNodes,
  getSelectedCount,
  getSelectedSum,
  getTestTooltipData,
  mapCellNodeValues,
} from '../util/interaction';
import { PivotSheet, SpreadSheet } from '@/sheet-type';
import type { S2DataConfig, S2Options } from '@/common/interface';

const s2Options: S2Options = {
  width: 600,
  height: 400,
  hierarchyType: 'grid',
};

describe('SpreadSheet Custom Grid Tests', () => {
  let s2: SpreadSheet;

  const baseDataConfig: Pick<S2DataConfig, 'data' | 'meta'> = {
    data: CustomGridData,
    meta: [
      {
        field: 'type',
        name: '类型',
      },
      {
        field: 'sub_type',
        name: '子类型',
      },
      {
        field: 'a-1',
        name: '层级1',
      },
      {
        field: 'a-1-1',
        name: '层级2',
      },
      {
        field: 'measure-1',
        name: '层级3',
      },
    ],
  };

  describe('Custom Row Grid Tests', () => {
    const customRowDataCfg: S2DataConfig = {
      ...baseDataConfig,
      fields: customRowGridSimpleFields,
    };

    beforeEach(async () => {
      s2 = new PivotSheet(getContainer(), customRowDataCfg, s2Options);
      await s2.render();
    });

    afterEach(() => {
      s2.destroy();
    });

    test('should disable valueInCols', () => {
      expect(s2.dataSet.fields.valueInCols).toBeFalsy();
    });

    test('should use custom grid pivot dataSet', () => {
      expect(s2.dataSet).toBeInstanceOf(CustomGridPivotDataSet);
    });

    test('should get correctly dataset fields', () => {
      expect(s2.dataSet.fields).toMatchSnapshot();
    });

    test('should render custom layout row nodes', () => {
      const rowNodes = s2.facet.getRowNodes().map((node) => {
        return {
          value: node.value,
          width: node.width,
          height: node.height,
          description: node.extra?.['description'],
        };
      });

      expect(rowNodes).toMatchSnapshot();
    });

    test('should calc correctly row index of leaf nodes', () => {
      const rowLeafNodes = s2.facet.getRowLeafNodes().map((node) => {
        return {
          value: node.value,
          rowIndex: node.rowIndex,
        };
      });

      expect(rowLeafNodes).toMatchSnapshot();
    });

    test('should calc correctly leaf nodes width after row resized', async () => {
      s2.setOptions({
        style: {
          rowCell: {
            widthByField: {
              'a-1-1': 40,
            },
          },
        },
      });
      await s2.render(false);

      const rowLeafNodes = s2.facet.getRowLeafNodes().map((node) => {
        return {
          field: node.field,
          width: node.width,
        };
      });

      const colLeafNodes = s2.facet.getColLeafNodes().map((node) => {
        return {
          field: node.field,
          width: node.width,
        };
      });

      expect(rowLeafNodes).toMatchSnapshot();
      expect(colLeafNodes).toMatchSnapshot();
    });

    test('should select custom row header cell', () => {
      // a-1
      const rowNode = s2.facet.getRowNodes()[0];

      // 选中 a-1
      s2.interaction.selectHeaderCell({
        cell: rowNode.belongsCell!,
      });

      // 选中单元格本身
      expect(s2.interaction.getActiveCells()).toHaveLength(1);
      // 高亮子节点
      expectHighlightActiveNodes(s2, [
        'root[&]a-1[&]a-1-1',
        'root[&]a-1[&]a-1-1[&]measure-1',
        'root[&]a-1[&]a-1-1[&]measure-2',
        'root[&]a-1[&]a-1-2',
      ]);

      // 取消选中 a - 1
      s2.interaction.selectHeaderCell({
        cell: rowNode.belongsCell!,
      });
      expect(s2.interaction.getActiveCells()).toBeEmpty();
    });

    test.each([
      { field: 'a-1', count: 6, sum: 34 },
      { field: 'a-1-1', count: 4, sum: 34 },
      { field: 'measure-1', count: 2, sum: 24 },
      { field: 'measure-2', count: 2, sum: 10 },
      { field: 'a-1-2', count: 2, sum: null },
      { field: 'a-1-2', count: 2, sum: null },
    ])(
      'should get selected cell summary infos for %o',
      ({ field, count, sum }) => {
        const rowNode = s2.facet
          .getRowNodes()
          .find((node) => node.field === field)!;

        // 选中
        s2.interaction.selectHeaderCell({
          cell: rowNode.belongsCell!,
        });

        const tooltipData = getTestTooltipData(s2, rowNode.belongsCell!);

        // 选中个数
        expect(getSelectedCount(tooltipData.summaries!)).toEqual(count);
        // 汇总数据总和
        expect(getSelectedSum(tooltipData.summaries!)).toEqual(sum);
      },
    );

    test('should render custom format corner text', () => {
      const cornerCellLabels = s2.facet.getCornerCells().map((cell) => {
        const value = cell.getActualText();
        const meta = cell.getMeta();

        return {
          value,
          field: meta.field,
        };
      });

      expect(cornerCellLabels).toMatchSnapshot();
    });

    test('should format custom rows', async () => {
      s2.setDataCfg({
        ...customRowDataCfg,
        meta: [
          {
            field: 'measure-1',
            formatter: (value) => `#-${value}`,
          },
          {
            field: 'a-1',
            formatter: (value) => `%-${value}`,
          },
          {
            field: 'a-1-1',
            formatter: (value) => `@-${value}`,
          },
          {
            field: 'a-1-2',
            formatter: (value) => `&-${value}`,
          },
          {
            field: 'a-2',
            formatter: (value) => `*-${value}`,
          },
        ],
      });
      await s2.render();

      const { rowNodes, dataCellTexts } = mapCellNodeValues(s2);

      expect(rowNodes).toMatchSnapshot();
      expect(dataCellTexts).toMatchSnapshot();
    });

    test('should collapse all custom row node', async () => {
      s2.setOptions({
        hierarchyType: 'tree',
        style: {
          rowCell: {
            collapseAll: true,
          },
        },
      });

      await s2.render(false);

      expect(s2.facet.getLayoutResult().rowsHierarchy.height).toEqual(60);
      expect(
        s2.facet.getRowNodes().every((node) => node.isCollapsed),
      ).toBeTruthy();
    });
  });

  describe('Custom Col Grid Tests', () => {
    const mapColNodes = (nodes = s2.facet.getColNodes()) => {
      return nodes.map((node) => {
        return {
          ...pick(node, ['field', 'value', 'x', 'y', 'width', 'height']),
          description: node.extra?.['description'],
        };
      });
    };
    const customColDataCfg: S2DataConfig = {
      ...baseDataConfig,
      fields: customColGridSimpleFields,
    };

    beforeEach(async () => {
      s2 = new PivotSheet(getContainer(), customColDataCfg, s2Options);
      await s2.render();
    });

    afterEach(() => {
      s2.destroy();
    });

    test('should enable valueInCols', () => {
      expect(s2.dataCfg.fields.valueInCols).toBeTruthy();
      expect(s2.dataSet.fields.valueInCols).toBeTruthy();
    });

    test('should get correctly dataset fields', () => {
      expect(s2.dataSet.fields).toMatchSnapshot();
    });

    /*
     * test('should use custom grid pivot dataSet', () => {
     *   expect(s2.dataSet).toBeInstanceOf(CustomGridPivotDataSet);
     * });
     */

    test('should render custom layout column nodes', () => {
      const colNodes = mapColNodes();

      expect(colNodes).toMatchSnapshot();
    });

    test('should calc correctly col index of leaf nodes', () => {
      const colLeafNodes = s2.facet.getColLeafNodes().map((node) => {
        return {
          value: node.value,
          colIndex: node.colIndex,
        };
      });

      expect(colLeafNodes).toMatchSnapshot();
    });

    test('should calc correctly leaf nodes width after column resized', async () => {
      s2.setOptions({
        style: {
          colCell: {
            heightByField: {
              'a-1-1': 40,
            },
          },
        },
      });
      await s2.render(false);

      const colNodes = s2.facet.getColNodes().map((node) => {
        return {
          value: node.value,
          height: node.height,
        };
      });

      expect(colNodes).toMatchSnapshot();
    });

    test('should select custom col header cell', () => {
      // a-1
      const colNode = s2.facet.getColNodes()[0];

      // 选中 a-1
      s2.interaction.selectHeaderCell({
        cell: colNode.belongsCell!,
      });

      // 选中单元格本身
      expect(s2.interaction.getActiveCells()).toHaveLength(1);
      // 高亮子节点
      expectHighlightActiveNodes(s2, [
        'root[&]a-1[&]a-1-1',
        'root[&]a-1[&]a-1-1[&]measure-1',
        'root[&]a-1[&]a-1-1[&]measure-2',
        'root[&]a-1[&]a-1-2',
      ]);

      // 取消选中 a - 1
      s2.interaction.selectHeaderCell({
        cell: colNode.belongsCell!,
      });
      expect(s2.interaction.getActiveCells()).toBeEmpty();
    });

    test.each([
      { field: 'a-1', count: 6, sum: 34 },
      { field: 'a-1-1', count: 4, sum: 34 },
      { field: 'measure-1', count: 2, sum: 24 },
      { field: 'measure-2', count: 2, sum: 10 },
      { field: 'a-1-2', count: 2, sum: null },
      { field: 'a-1-2', count: 2, sum: null },
    ])(
      'should get selected cell summary infos for %o',
      ({ field, count, sum }) => {
        const colNode = s2.facet
          .getColNodes()
          .find((node) => node.field === field)!;

        // 选中
        s2.interaction.selectHeaderCell({
          cell: colNode.belongsCell!,
        });

        const tooltipData = getTestTooltipData(s2, colNode.belongsCell!);

        // 选中个数
        expect(getSelectedCount(tooltipData.summaries)).toEqual(count);
        // 汇总数据总和
        expect(getSelectedSum(tooltipData.summaries)).toEqual(sum);
      },
    );

    test('should render custom format corner text', () => {
      const cornerCellLabels = s2.facet.getCornerCells().map((cell) => {
        const value = cell.getActualText();
        const meta = cell.getMeta();

        return {
          value,
          field: meta.field,
        };
      });

      expect(cornerCellLabels).toMatchSnapshot();
    });

    test('should hide columns', async () => {
      const hiddenColumns = ['root[&]a-1[&]a-1-1[&]measure-2'];

      await waitForRender(s2, () => {
        s2.interaction.hideColumns(hiddenColumns);
      });

      const hiddenColumnsDetail = s2.store.get('hiddenColumnsDetail', []);
      const [measureDetail] = hiddenColumnsDetail;

      expect(s2.options.interaction?.hiddenColumnFields).toEqual(hiddenColumns);
      expect(s2.facet.getColNodes().map((node) => node.field)).toEqual([
        'a-1',
        'a-1-1',
        'measure-1',
        'a-1-2',
        'a-2',
      ]);
      expect(hiddenColumnsDetail).toHaveLength(1);
      expect(measureDetail.displaySiblingNode.prev?.field).toEqual('measure-1');
      expect(measureDetail.displaySiblingNode.next?.field).toEqual('a-1-2');
      expect(measureDetail.hideColumnNodes).toHaveLength(1);
      expect(measureDetail.hideColumnNodes[0].field).toEqual('measure-2');
    });

    // https://github.com/antvis/S2/issues/1979
    test('should render correctly resize group for custom column fields', async () => {
      s2.setTheme({
        resizeArea: {
          backgroundOpacity: 1,
        },
      });
      await s2.render(false);

      const groups = s2.facet.foregroundGroup.getElementById<Group>(
        KEY_GROUP_COL_RESIZE_AREA,
      );

      expect(groups?.childNodes.length).toEqual(8);
    });

    // https://github.com/antvis/S2/issues/2017
    test('should format custom columns', async () => {
      s2.setDataCfg({
        ...customColDataCfg,
        meta: [
          {
            field: 'measure-1',
            formatter: (value) => `#-${value}`,
          },
          {
            field: 'measure-2',
            formatter: (value) => `666-${value}`,
          },
          {
            field: 'a-1',
            formatter: (value) => `%-${value}`,
          },
        ],
      });
      await s2.render();

      const { colNodes, dataCellTexts } = mapCellNodeValues(s2);

      expect(colNodes).toMatchSnapshot();
      expect(dataCellTexts).toMatchSnapshot();

      // 列头数值名称不应该按 formatter 格式化
      const measureNodes = colNodes.filter((node) =>
        node.id.startsWith('measure-'),
      );

      expect(
        measureNodes.every((node) => node.actualText === node.value),
      ).toBeTruthy();
    });

    // https://github.com/antvis/S2/issues/2117
    test('should calc correctly leaf nodes height and corner nodes', async () => {
      s2.setDataCfg({
        ...customColDataCfg,
        fields: {
          ...customColGridSimpleFields,
          columns: [
            {
              field: 'a-0',
              title: '测试-0',
              children: [
                {
                  field: 'a-0-1',
                  title: '测试-0-1',
                },
                {
                  field: 'a-0-2',
                  title: '测试-0-2',
                },
              ],
            },
            ...customColGridSimpleFields.columns!,
          ],
        },
      });
      await s2.render();

      const colNodes = mapColNodes();
      const cornerNodes = mapColNodes(s2.facet.getCornerNodes());
      const node1 = s2.facet
        .getColNodes()
        .find(({ field }) => field === 'a-0-1');
      const node2 = s2.facet
        .getColNodes()
        .find(({ field }) => field === 'a-0-2');

      expect(node1?.height).toEqual(60);
      expect(node1?.height).toEqual(node2?.height);
      expect(colNodes).toMatchSnapshot();
      expect(cornerNodes).toMatchSnapshot();
    });

    // https://github.com/antvis/S2/issues/2019
    test('should not collapse custom col node after row node collapsed', async () => {
      s2.setOptions({
        hierarchyType: 'tree',
        style: {
          rowCell: {
            collapseAll: true,
          },
        },
      });

      await s2.render(false);

      expect(s2.facet.getLayoutResult().colsHierarchy.height).toEqual(90);
      expect(
        s2.facet.getColNodes().some((node) => node.isCollapsed),
      ).toBeFalsy();
    });
  });
});
