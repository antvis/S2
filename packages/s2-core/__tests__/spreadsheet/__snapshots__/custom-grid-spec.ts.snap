// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should calc correctly col index of leaf nodes 1`] = `
Array [
  Object {
    "colIndex": 0,
    "value": "指标1",
  },
  Object {
    "colIndex": 1,
    "value": "指标2",
  },
  Object {
    "colIndex": 2,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "colIndex": 3,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should calc correctly leaf nodes height and corner nodes 1`] = `
Array [
  Object {
    "description": undefined,
    "field": "a-0",
    "height": 30,
    "value": "测试-0",
    "width": 192,
    "x": 0,
    "y": 0,
  },
  Object {
    "description": undefined,
    "field": "a-0-1",
    "height": 60,
    "value": "测试-0-1",
    "width": 96,
    "x": 0,
    "y": 30,
  },
  Object {
    "description": undefined,
    "field": "a-0-2",
    "height": 60,
    "value": "测试-0-2",
    "width": 96,
    "x": 96,
    "y": 30,
  },
  Object {
    "description": "a-1 描述",
    "field": "a-1",
    "height": 30,
    "value": "自定义节点 a-1",
    "width": 288,
    "x": 192,
    "y": 0,
  },
  Object {
    "description": "a-1-1 描述",
    "field": "a-1-1",
    "height": 30,
    "value": "自定义节点 a-1-1",
    "width": 192,
    "x": 192,
    "y": 30,
  },
  Object {
    "description": "指标1描述",
    "field": "measure-1",
    "height": 30,
    "value": "指标1",
    "width": 96,
    "x": 192,
    "y": 60,
  },
  Object {
    "description": "指标2描述",
    "field": "measure-2",
    "height": 30,
    "value": "指标2",
    "width": 96,
    "x": 288,
    "y": 60,
  },
  Object {
    "description": "a-1-2 描述",
    "field": "a-1-2",
    "height": 60,
    "value": "自定义节点 a-1-2",
    "width": 96,
    "x": 384,
    "y": 30,
  },
  Object {
    "description": "a-2 描述",
    "field": "a-2",
    "height": 90,
    "value": "自定义节点 a-2",
    "width": 96,
    "x": 480,
    "y": 0,
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should calc correctly leaf nodes height and corner nodes 2`] = `
Array [
  Object {
    "description": undefined,
    "field": "type",
    "height": 30,
    "value": "类型",
    "width": 96,
    "x": 0,
    "y": 60,
  },
  Object {
    "description": undefined,
    "field": "sub_type",
    "height": 30,
    "value": "子类型",
    "width": 96,
    "x": 96,
    "y": 60,
  },
  Object {
    "description": undefined,
    "field": "a-1",
    "height": 30,
    "value": "层级1",
    "width": 192,
    "x": 0,
    "y": 0,
  },
  Object {
    "description": undefined,
    "field": "a-1-1",
    "height": 30,
    "value": "层级2",
    "width": 192,
    "x": 0,
    "y": 30,
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should calc correctly leaf nodes width after column resized 1`] = `
Array [
  Object {
    "height": 30,
    "value": "自定义节点 a-1",
  },
  Object {
    "height": 40,
    "value": "自定义节点 a-1-1",
  },
  Object {
    "height": 30,
    "value": "指标1",
  },
  Object {
    "height": 30,
    "value": "指标2",
  },
  Object {
    "height": 70,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "height": 100,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should format custom columns 1`] = `
Array [
  Object {
    "actualText": "%-自定义节点 a-1",
    "id": "a-1",
    "value": "自定义节点 a-1",
  },
  Object {
    "actualText": "自定义节点 a-1-1",
    "id": "a-1-1",
    "value": "自定义节点 a-1-1",
  },
  Object {
    "actualText": "指标1",
    "id": "measure-1",
    "value": "指标1",
  },
  Object {
    "actualText": "指标2",
    "id": "measure-2",
    "value": "指标2",
  },
  Object {
    "actualText": "自定义节点 a...",
    "id": "a-1-2",
    "value": "自定义节点 a-1-2",
  },
  Object {
    "actualText": "自定义节点 a-2",
    "id": "a-2",
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should format custom columns 2`] = `
Array [
  "#-13",
  "#-11",
  "666-2",
  "666-8",
  "-",
  "-",
  "-",
  "-",
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should render custom format corner text 1`] = `
Array [
  Object {
    "field": "type",
    "value": "类型",
  },
  Object {
    "field": "sub_type",
    "value": "子类型",
  },
  Object {
    "field": "a-1",
    "value": "层级1",
  },
  Object {
    "field": "a-1-1",
    "value": "层级2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Col Grid Tests should render custom layout column nodes 1`] = `
Array [
  Object {
    "description": "a-1 描述",
    "field": "a-1",
    "height": 30,
    "value": "自定义节点 a-1",
    "width": 299.01,
    "x": 0,
    "y": 0,
  },
  Object {
    "description": "a-1-1 描述",
    "field": "a-1-1",
    "height": 30,
    "value": "自定义节点 a-1-1",
    "width": 199.34,
    "x": 0,
    "y": 30,
  },
  Object {
    "description": "指标1描述",
    "field": "measure-1",
    "height": 30,
    "value": "指标1",
    "width": 99.67,
    "x": 0,
    "y": 60,
  },
  Object {
    "description": "指标2描述",
    "field": "measure-2",
    "height": 30,
    "value": "指标2",
    "width": 99.67,
    "x": 99.67,
    "y": 60,
  },
  Object {
    "description": "a-1-2 描述",
    "field": "a-1-2",
    "height": 60,
    "value": "自定义节点 a-1-2",
    "width": 99.67,
    "x": 199.34,
    "y": 30,
  },
  Object {
    "description": "a-2 描述",
    "field": "a-2",
    "height": 90,
    "value": "自定义节点 a-2",
    "width": 99.67,
    "x": 299.01,
    "y": 0,
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should calc correctly leaf nodes width after row resized 1`] = `
Array [
  Object {
    "field": "measure-1",
    "width": 119.6,
  },
  Object {
    "field": "measure-2",
    "width": 119.6,
  },
  Object {
    "field": "a-1-2",
    "width": 159.6,
  },
  Object {
    "field": "a-2",
    "width": 279.2,
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should calc correctly leaf nodes width after row resized 2`] = `
Array [
  Object {
    "field": "sub_type",
    "width": 159.4,
  },
  Object {
    "field": "sub_type",
    "width": 159.4,
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should calc correctly row index of leaf nodes 1`] = `
Array [
  Object {
    "rowIndex": 0,
    "value": "指标1",
  },
  Object {
    "rowIndex": 1,
    "value": "指标2",
  },
  Object {
    "rowIndex": 2,
    "value": "自定义节点 a-1-2",
  },
  Object {
    "rowIndex": 3,
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should format custom rows 1`] = `
Array [
  Object {
    "actualText": "%-自定义节点 a-1",
    "id": "a-1",
    "value": "自定义节点 a-1",
  },
  Object {
    "actualText": "@-自定义节点 a-...",
    "id": "a-1-1",
    "value": "自定义节点 a-1-1",
  },
  Object {
    "actualText": "#-指标1",
    "id": "measure-1",
    "value": "指标1",
  },
  Object {
    "actualText": "指标2",
    "id": "measure-2",
    "value": "指标2",
  },
  Object {
    "actualText": "&-自定义节点 a-1-2",
    "id": "a-1-2",
    "value": "自定义节点 a-1-2",
  },
  Object {
    "actualText": "*-自定义节点 a-2",
    "id": "a-2",
    "value": "自定义节点 a-2",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should format custom rows 2`] = `
Array [
  "#-13",
  "2",
  "&-null",
  "*-null",
  "#-11",
  "8",
  "&-null",
  "*-null",
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should render custom format corner text 1`] = `
Array [
  Object {
    "field": "a-1",
    "value": "层级1",
  },
  Object {
    "field": "a-1-1",
    "value": "层级2",
  },
  Object {
    "field": "measure-1",
    "value": "层级3",
  },
  Object {
    "field": "type",
    "value": "类型",
  },
]
`;

exports[`SpreadSheet Custom Grid Tests Custom Row Grid Tests should render custom layout row nodes 1`] = `
Array [
  Object {
    "description": "a-1 描述",
    "height": 90,
    "value": "自定义节点 a-1",
    "width": 119.6,
  },
  Object {
    "description": "a-1-1 描述",
    "height": 60,
    "value": "自定义节点 a-1-1",
    "width": 119.6,
  },
  Object {
    "description": "指标1描述",
    "height": 30,
    "value": "指标1",
    "width": 119.6,
  },
  Object {
    "description": "指标2描述",
    "height": 30,
    "value": "指标2",
    "width": 119.6,
  },
  Object {
    "description": "a-1-2 描述",
    "height": 30,
    "value": "自定义节点 a-1-2",
    "width": 239.2,
  },
  Object {
    "description": "a-2 描述",
    "height": 30,
    "value": "自定义节点 a-2",
    "width": 358.79999999999995,
  },
]
`;
